torch/distributed/pipelining/stage.py

```diff
@@ -629,6 +637,7 @@ def perform_backward(
                     )
                 )
                 result = perform_backward(backward_type)()
+                perform_pp_grad_scaling()
```

Comment:
@fegin I think for DDP this won't work, becuase there is no way for me to insert pp_grad_scaling() call after backward and before DDP allreduce.

Actually, i'm not sure if this should matter.  I think it matters for FSDP becuase of how FSDP accumulates.

Reply:
Sorry, why does division before vs. after all-reduce/reduce-scatter matter?

Reply:
I first experimentally found that doing scaling after running fsdp's post_backward did not work.  (it led to the PP test failing numerics).  I need to study the FSDP code.  I thought that FSDP was doing some accumulation trick and perhaps moving the data to a different buffer hence the order mattering.  Now that I write this, that doesn't make sense to me.  The p.grad field should be the source of truth both before and after the reduce_scatter, right?  (I was probably confusing myself with the case where fsdp accumulates into .main_grad, but that's not relevant here.)

Reply:
Aside from floating-point differences, I am not sure why doing the scaling after reduce-scatter would be an issue 🤔 
The `sharded_param.grad` should be the source of truth.

Is this test running with mixed precision? If not, then there is no separate tensor for accumulation. Even if so, the scaling is an in-place op, so I am not sure how there could be any issue here / I do not have any good sense of what could be happening :/

Reply:
> so, in pp stage code, wouldself.submod.parameters.foo would refer to an unsharded param or a sharded param? I think that's a trick question, the answer being "it depends whether you look at it before or after calling post_backward". Before post_backward, i'd be scaling the 'unsharded' parameter. After post_backward, i'd be scaling the sharded parameter. Is that right?

I agree that it depends on before or after the last post-backward, but what does not make sense is why before or after matters. Both should be mathematically equivalent.

Reply:
yea. i think mathematically they should be equivalent.  I confirmed that if I apply some horrendous scaling like `p.grad /= 1000000`, it only seems to impact the test if I do it before post_backward.  If I do it after post_backward, the test behaves the same as if I do no scaling at all.  So I am not thinking about this as a numerics issue, but a mechanical one.  Probably somehow I'm modifying a wrong tensor.

Reply:
ok. in conclusion, the before/after thing was not relevant for numerics per-se.  I even did a simulation in the reference model where I scaled by 8 first and 2 later, vs 2 first and 8 later, and as expected this made no difference.  The only issues in the end were
 - PP/FSDP sync bug (prev PR https://github.com/pytorch/pytorch/pull/144535) made it so scaling after was racy
 - mixed-precision stores grads differently, making it so scaling before is more complicated (and leads to  missing p.grad errors if done naively)
 
So after fixing the race issue, the scaling implementation is to always scale after, which is also simple for mixed-precision.